src/Architecture.cpp

```diff
@@ -185,20 +189,43 @@ Architecture::Architecture(const std::uint16_t nQ, const CouplingMap& cm,
 }
 
 void Architecture::createDistanceTable() {
+  Matrix<double> edgeWeights(nqubits, std::vector<double>(nqubits, INFINITY));
   for (const auto& edge : couplingMap) {
+    edgeWeights.at(edge.first).at(edge.second) = COST_BIDIRECTIONAL_SWAP;
     if (couplingMap.find({edge.second, edge.first}) == couplingMap.end()) {
-      isBidirectional = false;
-      break;
+      edgeWeights.at(edge.second).at(edge.first) = COST_UNIDIRECTIONAL_SWAP;
     }
   }
 
-  if (isBidirectional) {
-    Dijkstra::buildTable(nqubits, couplingMap, distanceTable,
-                         Architecture::costHeuristicBidirectional);
-  } else {
-    Dijkstra::buildTable(nqubits, couplingMap, distanceTable,
-                         Architecture::costHeuristicUnidirectional);
+  Dijkstra::buildTable(nqubits, couplingMap, distanceTable, edgeWeights,
+                       Architecture::dijkstraNodeToCostNonFidelity);
```

Comment:
Why is there no distinction between directionality here anymore?

Reply:
This is actually interesting and also caught me a little by surprise at first. 

Currently (on the main git branch) the dijkstra implementation just counts the minimum number of edges from one qubit to another without regard of edge directionality (in extreme cases this could btw even lead to a wrong minimal path, as it could in fact be cheaper to take one more forward edge instead of a pure back-edge path, resulting in an overestimation of the cost). 
It then takes the number of `(steps-1)*COST_BIDIRECTIONAL_SWAP` resp. `(steps-1)*COST_UNIDIRECTIONAL_SWAP` (which again could break admissibility by overestimation in the unidirectional case, if I'm not mistaken). In case the last edge was a back-edge in the unidirectional case an additional `COST_DIRECTION_REVERSE` is added.
From this I gathered that the current implementation places the actual CNOT gate on the last edge and already accounts for a potential reversal of that gate.

In my implementation dijkstra is supplied with the actual cost of each swap via edge weights (these are set appropriately in the bidirectional and unidirectional case respectively). The resulting cost of dijkstra is therefore already the full swap cost. However, since dijkstra again finds only the full path and we want to place our CNOT on the last edge, we need to subtract the last swap:
In the bidirectional case (or on a forward edge): `cost - COST_BIDIRECTIONAL_SWAP`
On a back-edge: `cost - COST_UNIDIRECTIONAL_SWAP`. Since it is a back-edge, however, we need to add `COST_DIRECTION_REVERSE` again, resulting in `cost - COST_UNIDIRECTIONAL_SWAP + COST_DIRECTION_REVERSE` which is of course equivalent to `cost - COST_BIDIRECTIONAL_SWAP`
Therefore removing the last swap is equivalent in both cases (as we only want to remove the CNOTs of the swap), and the directionality of all edges in-between is accounted for (correctly) inside `Dijkstra::dijkstra`.

Reply:
Nice catch and thanks for the analysis! Now that makes a lot more sense. It need not be that detailed, but maybe a small comment about this would not hurt.

Reply:
I'm very sorry, I just realized a mistake in my reasoning there. In the unidirectional case with a forward-edge in the last step the above would actually not hold, as in this case it would still be necessary to apply a unidirectional swap, but if just executing one CNOT in the direction of the edge, we would not need the CNOT reversal. Therefore we would have `cost - COST_UNIDIRECTIONAL_SWAP`.
To solve this confusion once and for all, I now just introduced a new field in the Dijkstra nodes, which keeps track of the cost without the last swap. Of course this could also have been solved by looking at `Architecture::bidirectional()` similar to the (still inaccurate) solution before, but seeing as we just want the swap cost without the last swap it seems easier to just not add the last swap cost in the first place instead of inferring what it was and subtracting it after the fact.

Reply:
Alright. Sounds like a good solution to me! I'll look at it in detail later on.